torchqdynamics/odeint.py

```diff
@@ -83,35 +90,24 @@ def _fixed_odeint(qsolver, y0, t_save, dt, exp_ops, save_states):
         y_save = torch.zeros(len(t_save), *y0.shape).to(y0)
 
     if len(exp_ops) > 0:
-        exp_save = torch.zeros(len(t_save), len(exp_ops)).to(
-            device=y0.get_device(), dtype=torch.float
-        )
-
-    # save first step
-    save_counter = 0
-    if t_save[0] == 0.0:
-        if save_states:
-            y_save[0] = y0
-        for j, op in enumerate(exp_ops):
-            exp_save[save_counter, j] = torch.trace(op @ y0)
-        save_counter += 1
+        exp_save = torch.zeros(len(exp_ops),
+                               len(t_save)).to(device=y0.device, dtype=torch.float)
 
     # run the ode routine
     y = y0
-    t, t_max = 0, max(t_save)
-    while t < t_max:
-        # iterate solution
-        y = qsolver.forward(t, dt, y)
-
+    times = torch.arange(0.0, t_save[-1] + dt, dt)
+    save_counter = 0
+    for t in tqdm(times):
         # save solution
         if t >= t_save[save_counter]:
             if save_states:
                 y_save[save_counter] = y
-
             for j, op in enumerate(exp_ops):
-                exp_save[save_counter, j] = torch.trace(op @ y)
+                exp_save[j, save_counter] = torch.trace(op @ y).real
             save_counter += 1
-        t += dt
+
+        # iterate solution
+        y = qsolver.forward(t, dt, y)
```

Comment:
The loop is iterated one too many times. I understand you did that to not bother with saving the solution one more time, but I think it's better not to do that (`forward` could be slow, it could diverge after the final time step, or whatever).

So I would either go back to saving at `t=0.0` before the loop, do an automatic saving of `t=tsave[-1]` after the loop (my preferred option), or add a `break` inside the loop when the final time step is reached.

Reply:
100% agree